### ISSUE TITLE
Sampler / Loader on GPU

### ISSUE THREAD

**Issue by Foisunt** [pyg-team/pytorch_geometric#7321]
### 🚀 The feature, motivation and pitch

It sounds a littel wierd, but I'm sampling from a dataset that actually fits in VRAM.

The reason is that I try new aggregation methods and want to test the effect of different neighborhoods and batch sizes.
For some other models I also compute the adjacency matrix (that does not fit in VRAM when using the whole graph) to do some operations on it, this would also benefit from less overhead when sampling.

Right now I have the data on the GPU, send it back to the cpu to sample and then send the sampeld batches back to the GPU to actually do training/eval which is quite ugly.

### Additional context

Basically, fix this error ;)

NotImplementedError: Could not run 'pyg::neighbor_sample' with arguments from the 'CUDA' backend. [...]
pyg::neighbor_sample' is only available for these backends: [CPU, BackendSelect, Python, FuncTorchDynamicLayerBackMode, Functionalize, Named, Conjugate, Negative, ZeroTensor, ADInplaceOrView, AutogradOther, AutogradCPU, AutogradCUDA, AutogradXLA, AutogradMPS, AutogradXPU, AutogradHPU, AutogradLazy, AutogradMeta, Tracer, AutocastCPU, AutocastCUDA, FuncTorchBatched, FuncTorchVmapMode, Batched, VmapMode, FuncTorchGradWrapper, PythonTLSSnapshot, FuncTorchDynamicLayerFrontMode, PythonDispatcher]

**Comment by rusty1s** [comment]
We are working with NVIDIA team to bring in GPU-based sampling, but I cannot promise any release dates for that. I think holding features on the GPU, and just putting `edge_index` to CPU is pretty fast already, and lets you also use multiple workers to perform the sampling.
